### ISSUE TITLE
New lint rules for Strength Reduction optimizations

### ISSUE THREAD

**Issue by tdulcet** [astral-sh/ruff#14636]
Please consider adding new lint rules for [strength reduction](https://en.wikipedia.org/wiki/Strength_reduction) optimizations. All of the types of optimizations included in the below examples would improve the performance of Python code, some significantly, especially when the numbers are large. Ruff is fast, so lets help make Python fast as well! A few of the examples would also improve the accuracy/precision of the expression. Most of them are language agnostic, but they are particularly important in Python, as there is of course typically no compiler to perform the optimizations automatically (at least in CPython). The indented ones are just additional examples of what the proposed new lint rules could fix.

### Power of 2 integer optimizations
```py
x * 2                        # ⇒ x + x or x << 1
    x * 8                    # ⇒ x << 3
x // 2                       # ⇒ x >> 1
    x // 8                   # ⇒ x >> 3
x % 2                        # ⇒ x & 1
    x % 8                    # ⇒ x & 7
2 ** x                       # ⇒ 1 << x
    pow(2, x)
    y * 2 ** x               # ⇒ y << x
    8 ** 3                   # ⇒ 8 << 6 or 1 << 9
```
### Exponentiation and Modular optimizations
```py
(x * x) % y                  # ⇒ pow(x, 2, y)
    (x ** 2) % y
    (x * x * x) % y          # ⇒ pow(x, 3, y)
    (x ** 3) % y
    pow(x, 3) % y
x ** 2                       # ⇒ x * x
    pow(x, 2)
pow(x, y)                    # ⇒ x ** y
q, r = x // y, x % y         # ⇒ q, r = divmod(x, y)
    q = x // y; r = x % y
len(bin(x)[2:])              # ⇒ x.bit_length() # Python 3.1+
    len(bin(x).lstrip("-0b"))
```
### Math library related, improves accuracy
```py
math.ceil(x / y)             # ⇒ -(x // -y) # int only
int(math.sqrt(x))            # ⇒ math.isqrt(x) # Python 3.8+
x ** 0.5                     # ⇒ math.sqrt(x)
    x ** (1 / 2)
    pow(x, 0.5)
    math.pow(x, 0.5)
x ** (1 / 3)                 # ⇒ math.cbrt(x) # Python 3.11+
    pow(x, 1 / 3)
    math.pow(x, 1 / 3)
math.e ** x                  # ⇒ math.exp(x)
    pow(math.e, x)
    math.pow(math.e, x)
2.0 ** x                     # ⇒ math.exp2(x) # float only # Python 3.11+
    pow(2.0, x)
    math.pow(2.0, x)
math.exp(x) - 1              # ⇒ math.expm1(x) # Python 3.2+
    (math.e ** x) - 1
    pow(math.e, x) - 1
    math.pow(math.e, x) - 1
x % y                        # ⇒ math.fmod(x, y) # float only
(x * y) + z                  # ⇒ math.fma(x, y, z) # float only # Python 3.13+
math.pi * 2                  # ⇒ math.tau # Python 3.6+   
```
### Mathematical identities

May be useful in combination with other autofixes to simplify expressions or when refactoring
```py
x * -1               # ⇒ -x
x / -1.0             # ⇒ -x # float only
x // -1              # ⇒ -x # int only
x * 0                # ⇒ 0
x * 1                # ⇒ x
x + 0                # ⇒ x
x - 0                # ⇒ x
x - -y               # ⇒ x + y
x + -y               # ⇒ x - y
+x                   # ⇒ x
x / 1.0              # ⇒ x # float only
x // 1               # ⇒ x # int only
x ** 0               # ⇒ 1
x ** 1               # ⇒ x
x / y / z            # ⇒ x * z / y # float only 
x / 0.0              # ⇒ ERROR!
x // 0               # ⇒ ERROR!
x % 0                # ⇒ ERROR!
```
### Opinionated, but improves performance
```py
x / 2.0              # ⇒ x * 0.5 # float only
# Precompute small constant expressions
60 * 60 * 24         # ⇒ 86400
    (1 << 32) - 1    # ⇒ 0xFFFFFFFF
    "0" * 8          # ⇒ "00000000"
10.0 ** 3.0          # ⇒ 1e3 # float only
    2.0 * 10.0 ** 3.0# ⇒ 2e3
```

All of these should also work with `PLR6104` (#8877), so for example `x * 2` above should also catch `x *= 2` and offer to replace it with `x <<= 1`. There are of course more examples, but these are some of the ones I see most frequently in code reviews. As noted above, some of these may require type inference, but I believe an unsafe autofix would still be acceptable in many cases where the chance of a false positive would be low.

I could see some of these proposed rules being added to the existing `UP`, `PL`, `PERF`, `FURB` and `RUF` categories, among others. Please let me know if you think I should file this request with one of those upstream linters instead.

For a real world example, see this Python function to calculate the Jacobi symbol [before](https://rosettacode.org/wiki/Jacobi_symbol#Python) and [after](https://github.com/tdulcet/AutoPrimeNet/blob/d92313f882ec6dc11e9d79bd55f99ecee2645258/gimps_status.py#L611-L630) manually applying several of these strength reduction optimizations, which approximately doubled the resulting performance.

**Comment by dhruvmanila** [comment]
Interesting! Thanks for writing this up with all the examples provided. This would surely bring Ruff closer to being a ["compiler"](https://docs.astral.sh/ruff/contributing/#compilation-pipeline) ;)

**Comment by Avasam** [comment]
"Power of 2 integer optimizations" I find a *lot* less readable. Bitshifts and masks are not intuitive unless you actively use them regularly.

I'd expect those to be their own rule for when performance really matters. I'm all for "free speed optimizations" even (especially) for slow languages like Python. But these hurt readability a lot imo.

Some require type information that Ruff may not have atm (int vs float).

Everything else I'd love to see anytime!

**Comment by tdulcet** [comment]
> I'd expect those to be their own rule for when performance really matters. I'm all for "free speed optimizations" even (especially) for slow languages like Python.

Yes, I imagine each of these categories would include at least one rule. Note that the resulting performance improvement from those power of 2 optimizations can be significant (this is with Python 3.12):
```
$ python3 -m timeit -s 'x = 136279841' '2 ** x'
1 loop, best of 5: 551 msec per loop
$ python3 -m timeit -s 'x = 136279841' '1 << x'
500 loops, best of 5: 741 usec per loop
```
That is **744×** faster (ms to us).
```
$ python3 -m timeit -s 'x = (1 << 136279841) - 1' 'x % 2'
5 loops, best of 5: 44.3 msec per loop
$ python3 -m timeit -s 'x = (1 << 136279841) - 1' 'x & 1'
10000000 loops, best of 5: 33.8 nsec per loop
```
That is **762,980×** faster (ms to ns)! 🤯

**Comment by kaddkaka** [comment]
> > I'd expect those to be their own rule for when performance really matters. I'm all for "free speed optimizations" even (especially) for slow languages like Python.
> 
> Yes, I imagine each of these categories would include at least one rule. Note that the resulting performance improvement from those power of 2 optimizations can be significant (this is with Python 3.12):
> 
> ```
> $ python3 -m timeit -s 'x = 136279841' '2 ** x'
> 1 loop, best of 5: 551 msec per loop
> $ python3 -m timeit -s 'x = 136279841' '1 << x'
> 500 loops, best of 5: 741 usec per loop
> ```
> 
> That is **744×** faster (ms to us).
> 
> ```
> $ python3 -m timeit -s 'x = (1 << 136279841) - 1' 'x % 2'
> 5 loops, best of 5: 44.3 msec per loop
> $ python3 -m timeit -s 'x = (1 << 136279841) - 1' 'x & 1'
> 10000000 loops, best of 5: 33.8 nsec per loop
> ```
> 
> That is **762,980×** faster (ms to ns)! 🤯

I'm not sure how fair this comparison is. Such large numbers must be very rare?

**Comment by tdulcet** [comment]
> I'm not sure how fair this comparison is. Such large numbers must be very rare?

I used this large number to make it less noisy and easier to benchmark. Specifically, it is the [largest known prime number](https://en.wikipedia.org/wiki/Largest_known_prime_number). However, it is still much faster even with smaller numbers:
```
$ python3 -m timeit -s 'x = 127' '2 ** x'
1000000 loops, best of 5: 226 nsec per loop
$ python3 -m timeit -s 'x = 127' '1 << x'
10000000 loops, best of 5: 34.2 nsec per loop
```
That is 6.8× faster.
